Filter internal kubernetes labels from Prometheus metrics

grobie opened the pull request:
Kubernetes sets a bunch of internal labels on each docker container in
order to recover information in case of loss of the pod information
after a crash. These labels are not meaningful for Prometheus monitoring
and will create new timeseries in case they change (like restartCount).

With @jimmidyson's recent change to prefix labels with `container_label_`, users would have seen `container_label_io_kubernetes_container_hash`. 

@fabxc @timstclair @vishh

grobie commented:
Fixes #1312

jimmidyson commented:
We need the pod name, container name, etc that come from these labels (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/labels.go#L20). Perhaps we can just whitelist those labels & ignore others?

grobie commented:
We don't rely on these internal labels, but use the standard Prometheus relabel configs to build these from the available meta labels https://prometheus.io/docs/operating/configuration/#<kubernetes_sd_config>. I assumed that is the preferred way, but I can also create more specific filter if you think that's better. 

The current resulting label names like `container_label_io_kubernetes_pod_namespace` are a bit too long to be useful anyways.

jimmidyson commented:
I'm not at a computer at all this week to check but would you be able to share a snippet from exported kubelet /metrics from a kubernetes build from master? Just want to check what the raw pod metrics labels look like.

jimmidyson commented:
I'm pretty sure this will affect pod discovery which uses the container labels from exported metrics for pod name, etc but as I'm out on PTO all well I can't check myself.

grobie commented:
I'd be surprised if the Prometheus Kubernetes discovery client would rely
on labels exported by cAdvisor. As far as I can see, it reads these labels
from the Kubernetes API
https://github.com/prometheus/prometheus/blob/master/retrieval/discovery/kubernetes/pod.go#L294-L304.
This PR won't touch that at all, it only reduces the additional label noise
generated by cAdvisor.

On Thu, Aug 18, 2016 at 3:29 AM Jimmi Dyson <notifications@github.com>
wrote:

> I'm pretty sure this will affect pod discovery which uses the container
> labels from exported metrics for pod name, etc but as I'm out on PTO all
> well I can't check myself.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/cadvisor/pull/1426#issuecomment-240645493>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaGm8qQZwvi3dsUsCT2As4LyclZYdks5qhAnfgaJpZM4JnFwR>
> .
>

fabxc commented on metrics/prometheus.go:
```diff
@@ -535,6 +536,10 @@ func (c *PrometheusCollector) Collect(ch chan<- prometheus.Metric) {
 	containerEnvPrefix   = "container_env_"
+	// kubernetesInternalLabelPrefix is the common prefix of all labels set by
+	// Kubernetes in order to store internal information on docker containers.
+	// See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/labels.go.
```
Fix this link to the commit hash?

grobie replied:
@fabxc done

jimmidyson commented:
Doh I'm not talking about pod discovery but pod cgroup metrics. Pretty sure in kubernetes master, pod/container names will only be exposed via these container labels linked above. Just need to see a dump of /metrics from a kubelet built from master to confirm. We cleaned up duplicate label names iirc, relying on these container labels instead of deriving them from the docker container name.

grobie commented:
@jimmidyson I understand now what you mean. Given Prometheus scrapes one cadvisor instance per node, it's not possible to use relabeling to attach pod name/namespace labels to the metrics. I'll rework the PR.

grobie commented:
Here is an example of one metric exported by cadvisor (slightly reformatted to increase visibility).

google/cadvisor@2ed7198 kubelet@v1.3.5
```
container_cpu_cfs_periods_total{
  container_label_io_kubernetes_container_hash="5af8c3b4",
  container_label_io_kubernetes_container_name="sync",
  container_label_io_kubernetes_container_restartCount="1",
  container_label_io_kubernetes_container_terminationMessagePath="/dev/termination-log",
  container_label_io_kubernetes_pod_name="popularsearches-web-3165456836-2bfey",
  container_label_io_kubernetes_pod_namespace="popularsearches",
  container_label_io_kubernetes_pod_terminationGracePeriod="30",
  container_label_io_kubernetes_pod_uid="6a291e48-47c4-11e6-84a4-c81f66bdf8bd",
  id="/docker/68e1f15353921f4d6d4d998fa7293306c4ac828d04d1284e410ddaa75cf8cf25",
  image="redacted.com/popularsearches:42-16-ba6bd88",
  name="k8s_sync.5af8c3b4_popularsearches-web-3165456836-2bfey_popularsearches_6a291e48-47c4-11e6-84a4-c81f66bdf8bd_c02d3775"
} 72819
```

timstclair commented:
ok to test

timstclair commented:
Thus far we've avoided leaking Kubernetes concepts into cAdvisor, so I'm concerned about the precedent this is setting. I think I still don't quite understand what the problem here is. It sounds to me like the bug is that Kubernetes is abusing container labels, and this shouldn't be cAdvisor's problem.

If we really do need the logic in cAdvisor, I think a better approach would be to accept a list of labels to ignore, and have the Kubelet pass the list in rather than hardcoding it in cAdvisor. I also think the pod name, namespace, and container name are important to the identity and _should_ remain on the container.

grobie commented:
Alright, I've reworked the implementation and added a whitelist, allowing `io.kubernetes.pod.name` and `io.kubernetes.pod.namespace` still to be applied to all container metrics. There is currently no other way to attach such labels to cadvisor metrics.

grobie/cadvisor@12e1d20 kubelet@v1.3.5
```
container_cpu_cfs_periods_total{
  container_label_io_kubernetes_pod_name="popularsearches-web-3165456836-2bfey",
  container_label_io_kubernetes_pod_namespace="popularsearches",
  id="/docker/68e1f15353921f4d6d4d998fa7293306c4ac828d04d1284e410ddaa75cf8cf25",
  image="redacted.com/popularsearches:42-16-ba6bd88",
  name="k8s_sync.5af8c3b4_popularsearches-web-3165456836-2bfey_popularsearches_6a291e48-47c4-11e6-84a4-c81f66bdf8bd_c02d3775"
} 73700
```

grobie commented:
@timstclair I completely agree with you on leaking Kubernetes knowledge into cAdvisor. See my very first comment on the issue which started this discussion: https://github.com/google/cadvisor/issues/1312#issuecomment-240599649

That being said, without fundamentally changing Kubernetes, I don't see another way to work around this for now. cAdvisor reads these labels directly from the Docker API, so there is no other way to filter them then adding the knowledge to cADvisor. 

The huge problems with these labels is that they generate new timeseries every time they change. Prometheus is designed under the assumption that a labelset uniquely identifies one entity. These changed labels (e.g. when the restartCount gets increased) result in incorrect aggregation results for example. Some of the now filtered labels are static, but just noise and make it a lot harder to work with the cadvisor metrics.

Briefly before your comment I changed the implementation similar to your suggestion. I'm using a whitelist instead of a blacklist though. Pod name, namespace as well as the container name remain on the generated metrics as you can see in the example above.

PTAL again @jimmidyson @fabxc

grobie commented:
Alright, I didn't read your comment carefully enough. I'll rework it once again so that cadvisor simply accepts a filter function which can be set by kubernetes.

timstclair commented:
Yes, that was my suggestion. Specifically, it should be passed to [this function](https://github.com/google/cadvisor/blob/2ed7198f77395ee9a172878a0a7ab92ab59a2cfd/http/handlers.go#L92).

I would suggest renaming `ContainerNameToLabelsFunc` to `ContainerLablesFunc`, and have it take a few more parameters:
```
type ContainerNameToLabelsFunc func(containerName string, containerLabels map[string]string, containerEnv map[string]string) map[string]string
```
(also, make the sanitizeLabelName func exported, so it can be reused by implementers of the above function).

Finally, I would move all the existing label logic into an else condition. I.e., if that function is set, just use the output of that. Otherwise, do what we do today.

WDYT?

grobie commented:
I like the approach, any objections to make it even more generic and pass in the container object? There are also labels set based on `container.Aliases` and `container.Spec.Image` for example.

I'd then do something like `if labelsFunc == nil { labelsFunc = defaultLabelsFunc }` in the constructor.

jimmidyson commented:
SGTM. ContainerNameToLabelsFunc isn't used in kubernetes any more iirc. It was a stop gap until kubernetes implemented proper docker labels for this stuff.

muradm commented:
Will this fix the same behavior when using Docker Composer? Or we need another PR to remove container_label_com_docker_composer_* ?

grobie commented:
This change did not alter the behavior around metric labels. By default, cAdvisor will continue to export all available Docker labels using a container_label_* prefix. Kubernetes embeds cAdvisor and we pass a custom containerLabelsFunc there to only set a handful of explicitly chosen labels. I'm not aware of how cAdvisor is usually used in combination with docker compose, but something similar could probably be done. I suspect we could alternatively also provide a flag to let users explicitly set the list of labels they want to attach.

muradm commented:
Last proposal, with configurable set of labels seems to be most reasonable. In addition to the list it could be a map, something like:

`[-container.labels <label_name_as_set_on_container[:label_name_as_should_be_exported]> ... ]`

where `label_name_as_should_be_exported` is optional, if it is needed to change the name of exported label.

That would be great. Shall we open an issue for that?

Thanks in advance

grobie commented:
I guess opening a new issue and discussing this separate topic there makes
sense in either way.

On Mon, Sep 5, 2016 at 7:07 AM Murad <notifications@github.com> wrote:

> Last proposal, with configurable set of labels seems to be most
> reasonable. In addition to the list it could be a map, something like:
>
> [-container.labels
> <label_name_as_set_on_container[:label_name_as_should_be_exported]> ... ]
>
> where label_name_as_should_be_exported is optional, if it is needed to
> change the name of exported label.
>
> That would be great. Shall we open an issue for that?
>
> Thanks in advance
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/cadvisor/pull/1426#issuecomment-244721185>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaH-zDaDWCK2gExa9JRMBNOcr6BI4ks5qm_fsgaJpZM4JnFwR>
> .
>
